sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +528,454 @@ def transform(self, X):
+            either be equal to max_neighbors or max_neighbors + 1.
+            If the calling method is transform(), then its value needs to be
+            equal to max_neighbors and if calling method is fit_transform()
+            then its value must be equal to max_neighbors + 1.
+        """
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed in data to be "
+                             "transformed.")
+
+        # Get fitted data and ensure correct dimension
+        fitted_X = self._fitted_neighbors._fit_X
+        if X.shape[1] != fitted_X.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed.")
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+        if not np.any(row_total_missing):
+            return X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. The "
+                "missing features in these rows are imputed with column means."
+                    .format(self.row_max_missing * 100))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = mask[~bad_rows]
+            row_total_missing = mask.sum(axis=1)
+        row_has_missing = row_total_missing.astype(np.bool)
+
+        if np.any(row_has_missing):
+            if self.use_complete:
+                # Initializations
+                # Mask for fitted_X
+                mask_fx = _get_mask(fitted_X, np.nan)
+
+                # Locate unique patterns
+                patterns, row_pat_idx = _unique(
+                    mask, return_inverse=True, axis=0)
+
+                # Get row idx for receivers (missing)
+                receiver_row_missing_index, _ = np.where(mask)
+
+                # For every pattern, index receivers and potential donors
+                for p in range(len(patterns)):
+                    if not np.any(patterns[p]):
+                        continue
+                    # receivers are those with pattern 'p'
+                    row_has_missing_pat = (row_pat_idx == p)
+
+                    # Donors have features missing in receivers available
+                    # The bitwise-AND captures if something is missing in both
+                    donor_row_idx = ~np.any(patterns[p] & mask_fx, axis=1)
```

Comment:
So you're saying something is not a valid donor if there are any features that both the candidate and recipient have masked?

Why can't
```
[1 1 1 1 1 1 1 1 1 NaN 1]
```
donate to
```
[1 1 1 1 1 1 1 1 1 NaN NaN]
```
?

Reply:
If I'm right that your logic is incorrect, you should add a non-regression test...

Reply:
Why can't

[1 1 1 1 1 1 1 1 1 NaN 1]

donate to

[1 1 1 1 1 1 1 1 1 NaN NaN]

Yeah, I thought about this too. But this can lead to a situation where no. of neighbors for a given receiver sample becomes different from the one specified by the user. i.e. then it might become something like number of neighbors per feature per sample?

Reply:
I don't get how. Could you give an example?

Reply:
0,0,0,0
1,1,1,1
2,2,2,2
3,3,3,3
4,4,4,4
5,n,5,5
6,n,6,n
7,7,7,7
where n=nan
Consider sample at row=6 with k=5

To impute [6, 1], the 5-NN of the row with feature-1 available are:

1,1,1,1
2,2,2,2
3,3,3,3
4,4,4,4
7,7,7,7

 
but to impute [6, 3], the 5-NN are:

2,2,2,2
3,3,3,3
4,4,4,4
5,n,5,5
7,7,7,7

because the nan in [5, 1] does not "matter" in this case.

So, depending on the feature, we end up with different neighbors for the same sample.

Reply:
Yes, but I think this is correct behaviour according to the algorithm description. Yes, it is more difficult computationally, or at least in numpy.